### ISSUE TITLE
Rate limiting (patch provided)

### ISSUE THREAD

**Issue by janpascal** [digint/btrbk#72]
Hi,

When using btrbk over a DSL upstream, I'd really like to limit its bandwidth usage. I've prepared a patch that adds the "rate_limit" configuration file option, and adds a "-L" parameter to "pv", which is already in the command pipeline for the --progress command line options.

Please find attached a patch against the current master branch.
[rate_limit.patch.txt](https://github.com/digint/btrbk/files/169774/rate_limit.patch.txt)

**Comment by digint** [comment]
Thanks for your contribution, this looks very promising!

Does this work both ways?

This looks obvious:
```
btrfs send <local_src> | pv -L xxx | ssh <remote_host> btrfs receive <remote_target>
```

But what about this (if `volume  ssh://<remote_src>` is configured):
```
ssh <remote_host> btrfs send <remote_src> | pv -L xxx | btrfs receive <local_target>
```
Now the rate limiting happens AFTER the SSH transfer. Should work, but I'm not sure how `pv` handles this. I'm afraid the pv(1) man page does not give much hints on this.

Did you check that?

**Comment by janpascal** [comment]
I didn't check in this case, but I've used pv all over the place in pipes with ssh involved and it never matters where you put it. 

Axel Burri <notifications@github.com> schreef op 11 maart 2016 20:48:59 CET:
>Thanks for your contribution, this looks very promising!
>
>Does this work both ways?
>
>This looks obvious:
>```
>btrfs send <local_src> | pv -L xxx | ssh <remote_host> btrfs receive
><remote_target>
>```
>
>But what about this (if `volume  ssh://<remote_src>` is configured):
>```
>ssh <remote_host> btrfs send <remote_src> | pv -L xxx | btrfs receive
><local_target>
>```
>Now the rate limiting happens AFTER the SSH transfer. Should work, but
>I'm not sure how `pv` handles this. I'm afraid the pv(1) man page does
>not give much hints on this.
>
>Did you check that?
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/digint/btrbk/issues/72#issuecomment-195519661

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by digint** [comment]
Ok, I will merge this soon (I'm in the middle of refactoring some data structures in btrbk, this might still take some more time...)

**Comment by digint** [comment]
implemented in: f9ca7504c8f9297c7d25a93ef9f7ca3810c64a9e
